presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PhysicalResourceOptimizer.java

```diff
@@ -0,0 +1,308 @@
+    {
+        return this.executorCount;
+    }
+
+    public int getHashPartitionCount()
+    {
+        return this.hashPartitionCount;
+    }
+
+    public boolean isValid()
+    {
+        return ((this.executorCount > 0) && (this.hashPartitionCount > 0));
+    }
+
+    @Override
+    public PlanNode optimize(PlanNode plan, Session session, TypeProvider types, PlanVariableAllocator variableAllocator, PlanNodeIdAllocator idAllocator, WarningCollector warningCollector)
+    {
+        if (!isResourceManagementComputationEnabled(session)) {
+            return plan;
+        }
+
+        AbstractMap.SimpleImmutableEntry<Double, Boolean> sourceStatistics = new CollectSourceStats(metadata, session).collectSourceStats(plan);
+
+        if (0 == sourceStatistics.getValue().compareTo(true)) {
+            log.warn(String.format("Source missing statistics, skipping automatic resource tuning."));
+            return plan;
+        }
+
+        double inputDataInBytes = sourceStatistics.getKey();
+        if ((0 >= inputDataInBytes) || (inputDataInBytes > Double.MAX_VALUE) || (Double.isNaN(inputDataInBytes))) {
+            log.warn(String.format("Failed to retrieve correct size, data read=%.2f, skipping automatic resource tuning.", inputDataInBytes));
+            return plan;
+        }
+
+        double inputDataInGB = inputDataInBytes / 1000000000;
+        QueryTypeByDuration queryTypeByDuration = getQueryTypeByDuration(session);
+        boolean isCurrentQueryHourPlus = false;
+        if (queryTypeByDuration.equals(QueryTypeByDuration.DISABLED)) {
+            isCurrentQueryHourPlus = (inputDataInGB > THRESHOLD_SIZE_FOR_LONGER_QUERIES) ? true : false;
+        }
+        else if (queryTypeByDuration.equals(QueryTypeByDuration.LONG)) {
+            isCurrentQueryHourPlus = true;
+        }
+        double multiplier = getInputSizeMutltiplierForRM(session);
+
+        if (inputDataInGB * multiplier < Double.MAX_VALUE) {
+            inputDataInGB = inputDataInGB * multiplier;
+        }
+
+        boolean isRMDIRCUOptimized = isDircuOptimizedRMEnabled(session);
+        this.executorCount = calculateExecutorCount(inputDataInGB, isCurrentQueryHourPlus, isRMDIRCUOptimized);
+        this.hashPartitionCount = calculateHashPartitionCount(executorCount, isCurrentQueryHourPlus, isRMDIRCUOptimized);
+
+        return plan;
+    }
+
+    private int calculateHashPartitionCount(int executorCount, boolean isCurrentQueryHourPlus, boolean isRMDIRCUOptimized)
+    {
+        if (isCurrentQueryHourPlus) {
+            return (isRMDIRCUOptimized) ? calculateDircuOptimizedPartitonCountForLongerQueries(executorCount) :
+                   calculateLatencyOptimizedPartitionCountForLongerQueries(executorCount);
+        }
+        else {
+            return (isRMDIRCUOptimized) ? calculateDircuOptimizedPartitonCountForShorterQueries(executorCount) :
+                    calculateLatencyOptimizedPartitionCountForShorterQueries(executorCount);
+        }
+    }
```

Comment:
There are multiple combinations that are happening here for tuning two params: hash_partition_count and executorCount. This is bit confusing without any context. Can you add a detailed comment explaining the algorithm with a table like view or tree diagram explaining how different decisions are taken in the algorithm. That would be very helpful in understand the underlying logic.